### PR TITLE
Issue #3348: add validation of try with resources to the LocalFinalVariableName check

### PR DESCRIPTION
**Pull request by Nimfadora**
Issue: #3348 fixed
Also, since I needed to add extra test cases for LocalFinalVariableNameCheck. Diff report: https://nimfadora.github.io/

### CODE REVIEW

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/ConstantNameCheckTest.java [hunk 1]
```diff
@@ -86,7 +86,7 @@ public void testDefault()
 
         final String[] expected = {
             "25:29: " + getCheckMessage(MSG_INVALID_PATTERN, "badConstant", pattern),
-            "142:30: " + getCheckMessage(MSG_INVALID_PATTERN, "BAD__NAME", pattern),
+            "139:30: " + getCheckMessage(MSG_INVALID_PATTERN, "BAD__NAME", pattern),
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/ConstantNameCheckTest.java]: We are not expecting changes to this file.
Changes for your issue should be in an input file that no other tests uses to avoid this conflict.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalVariableNameCheckTest.java [hunk 2]
```diff
@@ -48,9 +48,9 @@ public void testDefault()
 
         final String[] expected = {
             "119:13: " + getCheckMessage(MSG_INVALID_PATTERN, "ABC", pattern),
-            "130:18: " + getCheckMessage(MSG_INVALID_PATTERN, "I", pattern),
-            "132:20: " + getCheckMessage(MSG_INVALID_PATTERN, "InnerBlockVariable", pattern),
-            "207:21: " + getCheckMessage(MSG_INVALID_PATTERN, "O", pattern),
+            "127:18: " + getCheckMessage(MSG_INVALID_PATTERN, "I", pattern),
+            "129:20: " + getCheckMessage(MSG_INVALID_PATTERN, "InnerBlockVariable", pattern),
+            "204:21: " + getCheckMessage(MSG_INVALID_PATTERN, "O", pattern),
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalVariableNameCheckTest.java]: Same here.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/MemberNameCheckTest.java [hunk 3]
```diff
@@ -57,7 +57,7 @@ public void testSpecified()
 
         final String[] expected = {
             "35:17: " + getCheckMessage(MSG_INVALID_PATTERN, "badMember", pattern),
-            "224:17: " + getCheckMessage(MSG_INVALID_PATTERN, "someMember", pattern),
+            "221:17: " + getCheckMessage(MSG_INVALID_PATTERN, "someMember", pattern),
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/MemberNameCheckTest.java]: Same here.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/MethodNameCheckTest.java [hunk 4]
```diff
@@ -57,7 +57,7 @@ public void testDefault()
         final String pattern = "^[a-z][a-zA-Z0-9]*$";
 
         final String[] expected = {
-            "137:10: " + getCheckMessage(MSG_INVALID_PATTERN, "ALL_UPPERCASE_METHOD", pattern),
+            "134:10: " + getCheckMessage(MSG_INVALID_PATTERN, "ALL_UPPERCASE_METHOD", pattern),
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/MethodNameCheckTest.java]: Same here.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/InputSimple.java [hunk 5]
```diff
@@ -118,9 +118,6 @@ private void localVariables()
         int abc = 0;
         int ABC = 0;
 
-        // final decls
-        final int cde = 0;
-        final int CDE = 0;
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/InputSimple.java]: Why is this change here? Please revert it.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/InputInner.java [hunk 6]
```diff
@@ -42,7 +42,7 @@
     /** demonstrate bug in handling static final **/
     static Object sWeird2 = new Object();
     
-    /** demonstrate bug in local final variable */
+
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/InputInner.java]: Why is this change in this file? Please revert it.

**Comment by Nimfadora** [comment]
@rnveach I fixed review issues. TeamCity build failed because of #3616 or?
If yes, @romani could you assign to me some new issue?

**Comment by rnveach** [comment]
@Nimfadora Yes, TC failures can be ignored for now.

Please read over https://github.com/checkstyle/checkstyle/wiki/PR-rules and keep all commits as 1.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalFinalVariableNameCheckTest.java [hunk 7]
```diff
@@ -56,9 +58,11 @@ public void testDefault()
         final String pattern = "^[a-z][a-zA-Z0-9]*$";
 
         final String[] expected = {
-            "123:19: " + getCheckMessage(MSG_INVALID_PATTERN, "CDE", pattern),
+            "17:19: " + getCheckMessage(MSG_INVALID_PATTERN, "CDE", pattern),
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalFinalVariableNameCheckTest.java]: Again, we are not expecting changes to this test.
Changes for your issue should be in an input file that no other tests uses to avoid this conflict.
Since this feature will increase violations, we are expecting violations for old tests to stay the same or increase. We don't expect any to disappear which is what is happening to `123`.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalFinalVariableNameCheckTest.java [hunk 8]
```diff
@@ -71,9 +75,9 @@ public void testSet()
         final String pattern = "[A-Z]+";
 
         final String[] expected = {
-            "122:19: " + getCheckMessage(MSG_INVALID_PATTERN, "cde", pattern),
+            "16:19: " + getCheckMessage(MSG_INVALID_PATTERN, "cde", pattern),
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalFinalVariableNameCheckTest.java]: same here.

**Comment by romani** [comment]
@Nimfadora , please focus on you proposals for now , proposals deadline is Apr 3, I need to see your proposal draft first.
you can continue with issues resolving till Apr 23.

**Comment by Nimfadora** [comment]
@romani, ok, clear

**Comment by romani** [comment]
@Nimfadora , I do not see proposals from you in GSoC website, please do.
Please do 1-2-3 proposals to avoid conflicts with students of better ranking.

**Comment by Nimfadora** [comment]
@romani, "1-2-3" is a number of proposals or number of projects on the github page?

**Comment by Vladlis** [comment]
@Nimfadora , it is a number of proposals, and each proposal can mention only one project.
Please do not forget provide links to all your contributions and detailed plan of work.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalFinalVariableNameCheckTest.java (hunk 7 above)
Comment (Nimfadora) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalFinalVariableNameCheckTest.java]: I moved all test inputs concerning LocalFinalVariableNameCheck which is the target class of this issue to separate file as a part of a tendency in Issue: #3891. I don't understand why it is bad, as you even have separate task on moving inputs to separate files.

**Comment by romani** [comment]
@Nimfadora , please share drafts of you proposals for any projects now, we need drafts to provide feedback before final dead-line.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalFinalVariableNameCheckTest.java (hunk 7 above)
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalFinalVariableNameCheckTest.java]: Referenced issue is to move inputs as is, with very little changes in the file. Only changes allowed are package and class name.
If input was moved as is, line number violations wouldn't change.
As shown above they changed from 123 to 17, so it looks like we have 100 lines that were removed.

**Comment by rnveach** [comment]
@Nimfadora ping
Since we are out of proposal deadline, do you think you can finish this?

**Comment by Nimfadora** [comment]
@rnveach, yes, of course I will finish it. A bit busy at university, think that today or tomorrow spend some time on it.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalFinalVariableNameCheckTest.java (hunk 7 above)
Comment (Nimfadora) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalFinalVariableNameCheckTest.java]: ok. I will move it as is.

**Comment by rnveach** [comment]
@Nimfadora Are you doing ok here?

**Comment by rnveach** [comment]
@Nimfadora CI is still failing, please fix.
Make sure you rebase on latest master.

**Comment by Nimfadora** [comment]
@rnveach CI now seem to be fine.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localfinalvariablename/InputLocalFinalVariableName.java [hunk 9]
```diff
@@ -188,6 +188,27 @@ void veryLong()
...
+                new FileInputStream(fileName), StandardCharsets.UTF_8))) {
+        }
+        finally {
+
+        }
+    }
+
+    void method2() throws Exception {
+        final String fileName = "Test";
+        try (final BufferedReader BR = new BufferedReader(new InputStreamReader(
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localfinalvariablename/InputLocalFinalVariableName.java]: `method` and `method2` are the exact same case. What is the point of having the same case twice?
Please change `method2` to a good case that won't produce a violation.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localfinalvariablename/InputLocalFinalVariableName.java [hunk 10]
```diff
@@ -20,7 +21,6 @@
     // Contains trailing whitespace ->
 
     // Name format tests
-    //
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localfinalvariablename/InputLocalFinalVariableName.java]: undo superfluous change.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localfinalvariablename/InputLocalFinalVariableName.java [hunk 11]
```diff
@@ -144,7 +144,7 @@ void ALL_UPPERCASE_METHOD()
     // A very, very long line that is OK because it matches the regexp "^.*is OK.*regexp.*$"
     // long line that has a tab ->	<- and would be OK if tab counted as 1 char
     // tabs that count as one char because of their position ->	<-   ->	<-, OK
- 
+
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localfinalvariablename/InputLocalFinalVariableName.java]: same

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalFinalVariableNameCheckTest.java [hunk 12]
```diff
@@ -59,6 +59,8 @@ public void testDefault()
 
         final String[] expected = {
             "123:19: " + getCheckMessage(MSG_INVALID_PATTERN, "CDE", pattern),
+            "194:29: " + getCheckMessage(MSG_INVALID_PATTERN, "BR", pattern),
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalFinalVariableNameCheckTest.java]: See my comment at https://github.com/checkstyle/checkstyle/pull/4116#discussion_r108506898 which I asked you to change but you didn't.

> We are not expecting changes to this file.
Changes for your issue should be in an input file that no other tests uses to avoid this conflict.

File: src/xdocs/config_naming.xml [hunk 13]
```diff
@@ -608,7 +608,7 @@
       <subsection name="Description">
         <p>
           Validates identifiers for local, <code>final</code> variables, including
-          <code>catch</code> parameters.
+          <code>catch</code> and <code>try-with-resource</code> parameters.
```
Comment (rnveach) [inline src/xdocs/config_naming.xml]: `try-with-resource` isn't a parameter. `catch` is the parameter.
change to `<code>catch</code> parameters and resources in <code>try</code> statements.

Please also make similar change to top of `LocalFinalVariableNameCheck`'s class javadoc.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localfinalvariablename/InputLocalFinalVariableName.java (hunk 10 above)
Comment (Nimfadora) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localfinalvariablename/InputLocalFinalVariableName.java]: Done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localfinalvariablename/InputLocalFinalVariableName.java (hunk 11 above)
Comment (Nimfadora) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localfinalvariablename/InputLocalFinalVariableName.java]: Done

File: src/xdocs/config_naming.xml (hunk 13 above)
Comment (Nimfadora) [inline src/xdocs/config_naming.xml]: Done

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalFinalVariableNameCheckTest.java (hunk 12 above)
Comment (Nimfadora) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalFinalVariableNameCheckTest.java]: Done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localfinalvariablename/InputLocalFinalVariableName.java (hunk 9 above)
Comment (Nimfadora) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localfinalvariablename/InputLocalFinalVariableName.java]: Done. However, not sure, if the new case is good enough. Coudn't imagine anything better.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalFinalVariableNameCheckTest.java (hunk 8 above)
Comment (Nimfadora) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalFinalVariableNameCheckTest.java]: Done

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalFinalVariableNameCheckTest.java (hunk 7 above)
Comment (Nimfadora) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalFinalVariableNameCheckTest.java]: Done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/InputInner.java (hunk 6 above)
Comment (Nimfadora) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/InputInner.java]: Done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/InputSimple.java (hunk 5 above)
Comment (Nimfadora) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/InputSimple.java]: Done

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/MethodNameCheckTest.java (hunk 4 above)
Comment (Nimfadora) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/MethodNameCheckTest.java]: Done

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/MemberNameCheckTest.java (hunk 3 above)
Comment (Nimfadora) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/MemberNameCheckTest.java]: Done

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalVariableNameCheckTest.java (hunk 2 above)
Comment (Nimfadora) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalVariableNameCheckTest.java]: Done

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/ConstantNameCheckTest.java (hunk 1 above)
Comment (Nimfadora) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/ConstantNameCheckTest.java]: Done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localfinalvariablename/InputLocalFinalVariableName.java [hunk 14]
```diff
@@ -188,6 +188,26 @@ void veryLong()
           enough talk */
     }
 
+    // check for try-with-resources
+    void method() throws Exception {
+        final String fileName = "Test";
+        try (BufferedReader bR = new BufferedReader(new InputStreamReader(
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localfinalvariablename/InputLocalFinalVariableName.java]: @Nimfadora I think you are mis-understanding what I am asking here. We need good cases (no violations) and bad cases (violations) involving try-with-resources. They need to be in their own input file and not modifying an existing one.

Undo the changes to this file. Create a new file like `InputLocalFinalVariableNameTryResources.java`.
Add all your cases to that file and create a test in `LocalFinalVariableNameCheckTest` to run it.

The example above is a good case, no violations.
The 2 examples you had before were, which are now gone, are bad cases, violations.
All 3 should be in this new input file.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localfinalvariablename/InputLocalFinalVariableName.java [hunk 15]
```diff
@@ -188,6 +188,26 @@ void veryLong()
...
+                new FileInputStream(fileName),"UTF-8"))) {
+        } finally {
+
+        }
+    }
+
+    void method2() throws Exception {
+       new java.util.Comparator() {
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localfinalvariablename/InputLocalFinalVariableName.java]: this has nothing to do with try-with-resources so it can be removed.
If you need examples, see the first post in the main issue.

Comment (Nimfadora) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localfinalvariablename/InputLocalFinalVariableName.java]: Done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localfinalvariablename/InputLocalFinalVariableName.java (hunk 14 above)
Comment (Nimfadora) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localfinalvariablename/InputLocalFinalVariableName.java]: Oh... Now I see your point. I moved all input for try-with-resorces check to separate file. However, I needed to add a one more test, to use this input file. Hope, now it is ok

Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localfinalvariablename/InputLocalFinalVariableName.java]: >  needed to add a one more test, to use this input file. Hope, now it is ok

That is what I stated.

> Add all your cases to that file and create a test in LocalFinalVariableNameCheckTest to run it.

**Comment by Nimfadora** [comment]
@rnveach Could you please give me a hint, why Wercker build is failing? I do not have an access to any kind of its output. Locally everything seems to be fine...

**Comment by rnveach** [comment]
wercker failure is `Could not resolve host: github.com`. We have been having this issue alot lately.
We can't manually restart wercker, only an update to the commit can re-run it.

**Comment by Nimfadora** [comment]
@romani all checks have passed, could you please look at it?

**Comment by romani** [comment]
@Nimfadora , I did not found diff report in this PR, please generate it on all projects that we have in property file.

diff.groovy at  https://github.com/checkstyle/contribution/tree/master/checkstyle-tester , please read README for usages and how to share report.
Before sharing report with us - review each point yourself.

**Comment by romani** [comment]
@Nimfadora , ping.

**Comment by Nimfadora** [comment]
@romani, yes, I'm here. Sorry a lot of tasks at university. Think today will manage to produce diff report.

**Comment by Nimfadora** [comment]
@romani, diff report is here https://nimfadora.github.io/

**Comment by romani** [comment]
https://nimfadora.github.io/findbugs-with-excldues/index.html

ThrowsCount is what you are testing ?
Please update config to use your Check - https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/my_check.xml#L26

**Comment by Nimfadora** [comment]
@romani, renewed report

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localfinalvariablename/InputLocalFinalVariableNameTryResources.java [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+        }
+        finally {
+
+        }
+    }
+
+    void method3() throws Exception {
+        final String fileName = "Test";
+        try (final BufferedReader BR = new BufferedReader(new InputStreamReader(
+                new FileInputStream(fileName), StandardCharsets.UTF_8))) {
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localfinalvariablename/InputLocalFinalVariableNameTryResources.java]: please make test cases with multiple variables in TRY to make sure that you validate all of them, not only first.
you can grab examples of non simple code from your report.

Comment (Nimfadora) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localfinalvariablename/InputLocalFinalVariableNameTryResources.java]: @romani, I added new test cases but wercker build failing all the time, what's the point?

**Comment by Nimfadora** [comment]
@romani, I've added test cases, but what should I do with codecov checks?

**Comment by rnveach** [comment]
If you are referring to `codecov` CI, then it can be ignored as travis passes.
